master/internal/api_trials.go

```diff
@@ -531,6 +534,41 @@ func (a *apiServer) PostTrialProfilerMetricsBatch(
 	return &apiv1.PostTrialProfilerMetricsBatchResponse{}, errs.ErrorOrNil()
 }
 
+func (a *apiServer) TrialPreemptionSignal(
+	req *apiv1.TrialPreemptionSignalRequest,
+	resp apiv1.Determined_TrialPreemptionSignalServer,
+) error {
+	eID, rID, err := a.m.db.TrialExperimentAndRequestID(int(req.TrialId))
+	switch {
+	case err == db.ErrNotFound:
+		return trialNotFound
+	case err != nil:
+		return err
+	}
+	trialAddr := actor.Addr("experiments", eID, rID)
+
+	signal := make(chan struct{}, 1)
+	a.m.system.TellAt(trialAddr, trialWatchPreemption{signal: signal})
+	unwatch := func() { a.m.system.TellAt(trialAddr, trialUnwatchPreemption{signal: signal}) }
+
+	for {
+		select {
+		case _, ok := <-signal:
+			if !ok {
+				// Do not call unwatch() when the channel is closed.
```

Comment:
I think the cleanup process could be simplified by putting `unwatch` in a `defer`.  There's some design philosophy here, since what I'm asking for is technically unnecessary.

Any time you have objects with non-concentric lifetimes you have to deal with how things shut each other down.  The best discussion about this that I've ever seen is in the [GObject memory model](https://developer.gnome.org/gobject/stable/gobject-memory.html), especially the section on reference cycles.

My practical takeaway from that article is that shutdown events between objects that reference each other are inherently racy and it's much easier to reason about them if they are idempotent and always get called.

Or in other words, after studying the code as written, I can see that the only time the trial would close the signal is a case where unwatch would be pointless.  But an extra `unwatch` call in that case is a noop, so if you put `unwatch` in a defer statement, you can read `TrialPreemptSignal` and understand that it is correct without having to think through the asynchronous interactions between two different parts of the system.

Reply:
> My practical takeaway from that article is that shutdown events between objects that reference each other are inherently racy and it's much easier to reason about them if they are idempotent and always get called.

yeah this makes sense.

> Or in other words, after studying the code as written, I can see that the only time the trial would close the signal is a case where unwatch would be pointless. But an extra unwatch call in that case is a noop, so if you put unwatch in a defer statement, you can read TrialPreemptSignal and understand that it is correct without having to think through the asynchronous interactions between two different parts of the system.

yeah i can call in defer, for some reason i didn't realize a tell to a nil actor was a noop

Reply:
Oh, I guess I don't know if Telling a `nil` actor is safe, but it should never be `nil` if we get to the `defer` statement.  It might be closed, but shouldn't be nil.